<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="utf-8" />
    <title>Gibbs-Sampler
</title>
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <meta name="description" content="Algorithm for motive finding." />
    <meta name="author" content="Etschbeijer" />

    <script src="https://code.jquery.com/jquery-1.8.0.js"></script>
    <script src="https://code.jquery.com/ui/1.8.23/jquery-ui.js"></script>
    <script src="https://netdna.bootstrapcdn.com/twitter-bootstrap/2.2.1/js/bootstrap.min.js"></script>
    <link href="https://netdna.bootstrapcdn.com/twitter-bootstrap/2.2.1/css/bootstrap-combined.min.css" rel="stylesheet" />
    <meta http-equiv="X-UA-Compatible" content="IE=11">
    <script src="https://cdn.plot.ly/plotly-latest.min.js"></script>
    <link type="text/css" rel="stylesheet" href="/GibbsSampler/content/style.css" /> <!--/GibbsSampler-->
    <link type="text/css" rel="stylesheet" href="/GibbsSampler/styles/CSBStyles.css" /> <!--/GibbsSampler-->
    <script type="text/javascript" src="/GibbsSampler/content/tips.js"></script> <!--/GibbsSampler-->
    <!-- HTML5 shim, for IE6-8 support of HTML5 elements -->
    <!--[if lt IE 9]>
      <script src="https://oss.maxcdn.com/html5shiv/3.7.2/html5shiv.min.js"></script>
    <![endif]-->
</head>
  <body>
    <div class="container">
      <div class="masthead">
        <ul class="nav nav-pills pull-right">
          <li><a href="http://fsharp.org">fsharp.org</a></li>
          <li><a href="https://github.com/Etschbeijer/GibbsSampler">github page</a></li>
        </ul>
        <h3 class="muted"><a href="/GibbsSampler/index.html">GibbsSampler</a></h3>
      </div>
      <hr />
      <div class="row">
        <div class="span9" id="main">
          
<h1><a name="Gibbs-Sampler" class="anchor" href="#Gibbs-Sampler">Gibbs-Sampler</a></h1>
<h2><a name="motif-search-de-novo-or-with-prior-knowledge" class="anchor" href="#motif-search-de-novo-or-with-prior-knowledge">motif search de novo or with prior knowledge</a></h2>
<h5><a name="In-this-documentation-the-basic-principles-of-the-Gibbs-Sampling-will-be-explained-with-a-focus-on-its-use-in-computational-biology" class="anchor" href="#In-this-documentation-the-basic-principles-of-the-Gibbs-Sampling-will-be-explained-with-a-focus-on-its-use-in-computational-biology">In this documentation the basic principles of the Gibbs-Sampling will be explained with a focus on its use in computational biology.</a></h5>
<p>Also does this project build currently on top of <a href="https://csbiology.github.io/BioFSharp/">BioFSharp</a> and will be integrated at a later step.</p>
<p>Gibbs-Sampling is an algorithm based on the Markov-Chain-Monte-Carlo method, which enables the calculation of a joint distribution
solely based on conditional distributions without prior knowledge. In computational biology it is commonly used to find
conserved segments, so called motifs, that are part of different sequences. Motifs are segments of sequences with a similar structure and can be, for example,
transcription-factor-binding-sites in DNA or ligand-binding-sites in proteins. These binding-sites enable the binding of those elements, in order to regulate the activity of the target.
Transcription-factor-binding-sites are often located in the promoter and 5'UTR of genes, which enable the regulation of the gene activity downstreams.</p>
<p>My research-group researches the heath-shock-response of the alga Chlamydomonas reinhardtii by evaluating the proteome compositions of different mutants and
targeting the corresponding genes. The goal of implementing this algorithm is to detect the transcription-factor-binding-sites of Heat-Shock-Factors (HSFs), proteins that
regulate the gene-activity during heat-stress. These binding-sites are called Heat-Shock-Elements (HSEs) and it is suspected that they are located in many genes.
Sadly, they are poorly conserved compared to the HSEs of other organims and we hope that we can detect more HSE regulated genes by implementing a Gibbs-Sampling-Algorithm.</p>
<p>There are two distinct types of Gibbs-Sampling-Algorithms, the Site-Sampler and motif-Sampler. Both of them share the basic idea but while the Site-Sampler requires
a motif in every sequence to work properly, the motif-Sampler does not. The motif-Sampler can also find multiple motifs per sequence, while the Site-Sampler can only find one
but works faster and is more sensitive for less conserved structures then the motif-Sampler. In order for both to wok proberly, the length of the sequence must be known.
The core idea of the Site-Sampler will now be explained with an example:</p>
<p><img src="img/sequences1.png" alt="Figure 1" /></p>
<h4><a name="Figure-1" class="anchor" href="#Figure-1">Figure 1.</a></h4>
<p><img src="img/sequences2.png" alt="Figure 2" /></p>
<h4><a name="Figure-2" class="anchor" href="#Figure-2">Figure 2.</a></h4>
<p>You have four sequences (see Figure 1) and you pick one at random and put it aside. After that you choose one segment with the length of your motif in every sequence you
did not choose (see Figure 2).</p>
<p><img src="img/pfm.png" alt="Figure 3" /></p>
<h4><a name="Figure-3" class="anchor" href="#Figure-3">Figure 3.</a></h4>
<p><img src="img/ppm.png" alt="Figure 4" /></p>
<h4><a name="Figure-4" class="anchor" href="#Figure-4">Figure 4.</a></h4>
<p>Align these segments with eacht other and count the amount of elements at each position in order to create a so called Position-Frequency-Matrix (PFM) (see Figure 3).
In the next step ypu create a so called position probability matrix by adding a pseudocount to each element at each position and normalize by dividing each element at
each position through the sum of the column (see Figure 4). You add a pseudocount in order to avoid dividing through zero because often times the segments are rather short and do not
contain every element and it is still possible to encounter one in the rest of the sequence.</p>
<p><img src="img/cfv.png" alt="Figure 5" /></p>
<h4><a name="Figure-5" class="anchor" href="#Figure-5">Figure 5.</a></h4>
<p><img src="img/SegmentScore.png" alt="Figure 6" /></p>
<h4><a name="Figure-6" class="anchor" href="#Figure-6">Figure 6.</a></h4>
<p>Then you create a so called Frequency-Composite-Vector (FCV) by counting the amount of elements in the unchosen sequences without the elements that are part of the segments
that are used to create the PFM. In the following you work with the chosen sequence by increasing the counts of the vector with amount of elements in this sequence and add a pseudocount to each element.<br />
The PPM is now used to calculate the likelyhood of each segment to be a motif by multiplying the likelyhoods of the found elements to be at that position (see Figure 4 &amp; 6).</p>
<p><img src="img/pcm.png" alt="Figure 7" /></p>
<h4><a name="Figure-7" class="anchor" href="#Figure-7">Figure 7.</a></h4>
<p>Now that the segment likelyhoods are calculated it is time to check, how different they are from the background. For that, the FCV is tuned by creating a new FCV for every segment.
For that, the count of eacht element is substracted from the total counts of elements in the FCV but it must not be smaller than the pseudocount. This is done because some input sequences are
very small and significant element combinations (e.g. several Ts in a row) could get low weight because they are otherwise evenly distributed. So by reducing the count for the segment, their weight
is increased. After that the FCVs are normalized by dividing every element count by the the sum of the vector. These new vectors are Probability-Composite-Vectors (PCV) and represent the likelyhood to pick a
element by chance when picking a random element from the sequences (see Figure 7).</p>
<p><img src="img/pwms.png" alt="Figure 8" /></p>
<h4><a name="Figure-8" class="anchor" href="#Figure-8">Figure 8.</a></h4>
<p>The PCVs are then used to calculate the likelyhood for a segment to be background by multiplying the likelyhood of the element to be picked at random in the segment. To calculate this likelyhood only the
PCVs that were created without the elements of the segment are used and these new background segment scores are used to normalize the PPM segment score in the earlier step. The result of this division is then
log2 transformed to get the so called Position-Weight-Matrix-Score (PWMS). Based on the PWMSs in Figure 8, PWMS is more likely to be a conserved motif and would be remembered (see Figure 8).
A Position-Weight-Matrix (PWM) is a PPM normalized to its background but becasue we assume no prior knowledge about the background the vectors and the PWMs need to be calculatedfor every segment anew.</p>
<p>The motif with the best PWMS is remembered of the chosen sequence and then a new sequence is chosen at random and the calculation of the vectors and matrices repeated in the same fashion until every sequence
had been chosen. Then the remembered motifs are used to calculate a new PPM. A sequence is randomly chosen and the best PWMS is looked for, using the PPM. Then the PWMS of this iteration is compared with
the remembered PWMS and if it is better it replaces the old one and a new PPM is calculated. Otherwise the old PPM is kept and the next Sequence chosen at random and these iterations are repeated until convergence
occures. That means, that always the same motifs are chosen because they have the best PWMSs. This happens because by chance a random start will happen in one of the conserved structures and then it is more likely to
find another conserverd structure in one of the other sequences, because they are more similiar to each other than the background and have a higher PWMS that way.</p>
<p>The first step in analyzing your bio sequence is transforming it into an array of BioArrays of IBioItems. This can be done easily by applaying the BioArray.ofNucleotideString of the BioFSharp-Library.</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l"> 1: </span>
<span class="l"> 2: </span>
<span class="l"> 3: </span>
<span class="l"> 4: </span>
<span class="l"> 5: </span>
<span class="l"> 6: </span>
<span class="l"> 7: </span>
<span class="l"> 8: </span>
<span class="l"> 9: </span>
<span class="l">10: </span>
<span class="l">11: </span>
<span class="l">12: </span>
<span class="l">13: </span>
<span class="l">14: </span>
<span class="l">15: </span>
<span class="l">16: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="k">open</span> <span onmouseout="hideTip(event, 'fs1', 1)" onmouseover="showTip(event, 'fs1', 1)" class="id">BioFSharp</span>
<span class="k">open</span> <span onmouseout="hideTip(event, 'fs2', 2)" onmouseover="showTip(event, 'fs2', 2)" class="id">FSharpAux</span>
<span class="k">open</span> <span onmouseout="hideTip(event, 'fs3', 3)" onmouseover="showTip(event, 'fs3', 3)" class="id">GibbsSampling</span>
<span class="k">open</span> <span onmouseout="hideTip(event, 'fs4', 4)" onmouseover="showTip(event, 'fs4', 4)" class="id">SiteSampler</span>
<span class="k">open</span> <span onmouseout="hideTip(event, 'fs3', 5)" onmouseover="showTip(event, 'fs3', 5)" class="id">GibbsSampling</span>
<span class="k">open</span> <span onmouseout="hideTip(event, 'fs3', 6)" onmouseover="showTip(event, 'fs3', 6)" class="id">GibbsSampling</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs5', 7)" onmouseover="showTip(event, 'fs5', 7)" class="id">PositionMatrix</span>

<span class="k">let</span> <span onmouseout="hideTip(event, 'fs6', 8)" onmouseover="showTip(event, 'fs6', 8)" class="id">bioSources</span> <span class="o">=</span>
    <span class="pn">[</span>
        <span class="s">&quot;GTGGCTGCACCACGTGTATGC&quot;</span>
        <span class="s">&quot;ACATCGCATCACGTGACCAGT&quot;</span>
        <span class="s">&quot;CCTCGCACGTGGTGGTACAGT&quot;</span>
        <span class="s">&quot;CTCGTTAGGACCATCACGTGA&quot;</span>
    <span class="pn">]</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs7', 9)" onmouseover="showTip(event, 'fs7', 9)" class="m">Seq</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs8', 10)" onmouseover="showTip(event, 'fs8', 10)" class="id">map</span> <span class="pn">(</span><span class="k">fun</span> <span onmouseout="hideTip(event, 'fs9', 11)" onmouseover="showTip(event, 'fs9', 11)" class="id">item</span> <span class="k">-&gt;</span> <span onmouseout="hideTip(event, 'fs10', 12)" onmouseover="showTip(event, 'fs10', 12)" class="m">BioArray</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs11', 13)" onmouseover="showTip(event, 'fs11', 13)" class="id">ofNucleotideString</span> <span onmouseout="hideTip(event, 'fs9', 14)" onmouseover="showTip(event, 'fs9', 14)" class="id">item</span><span class="pn">)</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs12', 15)" onmouseover="showTip(event, 'fs12', 15)" class="m">Array</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs13', 16)" onmouseover="showTip(event, 'fs13', 16)" class="id">ofSeq</span>
</code></pre></td>
</tr>
</table>
<p>Based on the knwoledge different kinds of site sampling functions can be chosen. All depend on knowing the length of the motif you are looking for.
The most basic Site-Sampling function requires the motif length, pseudoCount, the sources in the form of an array of BioArrays of IBioItemsand the used alphabet.
The alphabet is a collection of bioitems representing DNA bases or amino acids that are part of your sequence and enable the avoidance of unnecessary
iterations, for ecample Uracil positions when the sequence consists of DNA or Amino-Acids.</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs14', 17)" onmouseover="showTip(event, 'fs14', 17)" class="id">alphabet</span> <span class="o">=</span> <span class="pn">[|</span><span onmouseout="hideTip(event, 'fs15', 18)" onmouseover="showTip(event, 'fs15', 18)" class="m">Nucleotides</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs16', 19)" onmouseover="showTip(event, 'fs16', 19)" class="rt">Nucleotide</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs17', 20)" onmouseover="showTip(event, 'fs17', 20)" class="id">A</span><span class="pn">;</span> <span onmouseout="hideTip(event, 'fs15', 21)" onmouseover="showTip(event, 'fs15', 21)" class="m">Nucleotides</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs16', 22)" onmouseover="showTip(event, 'fs16', 22)" class="rt">Nucleotide</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs18', 23)" onmouseover="showTip(event, 'fs18', 23)" class="id">T</span><span class="pn">;</span> <span onmouseout="hideTip(event, 'fs15', 24)" onmouseover="showTip(event, 'fs15', 24)" class="m">Nucleotides</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs16', 25)" onmouseover="showTip(event, 'fs16', 25)" class="rt">Nucleotide</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs19', 26)" onmouseover="showTip(event, 'fs19', 26)" class="id">G</span><span class="pn">;</span> <span onmouseout="hideTip(event, 'fs15', 27)" onmouseover="showTip(event, 'fs15', 27)" class="m">Nucleotides</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs16', 28)" onmouseover="showTip(event, 'fs16', 28)" class="rt">Nucleotide</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs20', 29)" onmouseover="showTip(event, 'fs20', 29)" class="id">C</span><span class="pn">|]</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs21', 30)" onmouseover="showTip(event, 'fs21', 30)" class="id">motiveLength</span> <span class="o">=</span> <span class="n">6</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs22', 31)" onmouseover="showTip(event, 'fs22', 31)" class="id">pseudoCount</span> <span class="o">=</span> <span class="n">0.0001</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs23', 32)" onmouseover="showTip(event, 'fs23', 32)" class="id">randomMotifs</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs24', 33)" onmouseover="showTip(event, 'fs24', 33)" class="fn">getPWMOfRandomStarts</span> <span onmouseout="hideTip(event, 'fs21', 34)" onmouseover="showTip(event, 'fs21', 34)" class="id">motiveLength</span> <span onmouseout="hideTip(event, 'fs22', 35)" onmouseover="showTip(event, 'fs22', 35)" class="id">pseudoCount</span> <span onmouseout="hideTip(event, 'fs14', 36)" onmouseover="showTip(event, 'fs14', 36)" class="id">alphabet</span> <span onmouseout="hideTip(event, 'fs6', 37)" onmouseover="showTip(event, 'fs6', 37)" class="id">bioSources</span>
</code></pre></td>
</tr>
</table>
<p>This function gives you the PWMSs and Positions of the best fitting motifs after the random initialization.</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs25', 38)" onmouseover="showTip(event, 'fs25', 38)" class="id">firstConvergedMotifs</span> <span class="o">=</span>
    <span onmouseout="hideTip(event, 'fs23', 39)" onmouseover="showTip(event, 'fs23', 39)" class="id">randomMotifs</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs26', 40)" onmouseover="showTip(event, 'fs26', 40)" class="fn">getBestPWMSsWithStartPositions</span> <span onmouseout="hideTip(event, 'fs21', 41)" onmouseover="showTip(event, 'fs21', 41)" class="id">motiveLength</span> <span onmouseout="hideTip(event, 'fs22', 42)" onmouseover="showTip(event, 'fs22', 42)" class="id">pseudoCount</span> <span onmouseout="hideTip(event, 'fs14', 43)" onmouseover="showTip(event, 'fs14', 43)" class="id">alphabet</span> <span onmouseout="hideTip(event, 'fs6', 44)" onmouseover="showTip(event, 'fs6', 44)" class="id">bioSources</span>
</code></pre></td>
</tr>
</table>
<p>This function iterates through the sequences until its first convergence and gives you then the PWMSs and Positions of the best motif of every sequence. It requires a
Startposition in contrast to the first function because it has no random initialization but it can be piped easily behind the other.</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs27', 45)" onmouseover="showTip(event, 'fs27', 45)" class="id">bestFittingMotifs</span> <span class="o">=</span>
    <span onmouseout="hideTip(event, 'fs25', 46)" onmouseover="showTip(event, 'fs25', 46)" class="id">firstConvergedMotifs</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs28', 47)" onmouseover="showTip(event, 'fs28', 47)" class="fn">getLeftShiftedBestPWMSs</span> <span onmouseout="hideTip(event, 'fs21', 48)" onmouseover="showTip(event, 'fs21', 48)" class="id">motiveLength</span> <span onmouseout="hideTip(event, 'fs22', 49)" onmouseover="showTip(event, 'fs22', 49)" class="id">pseudoCount</span> <span onmouseout="hideTip(event, 'fs14', 50)" onmouseover="showTip(event, 'fs14', 50)" class="id">alphabet</span> <span onmouseout="hideTip(event, 'fs6', 51)" onmouseover="showTip(event, 'fs6', 51)" class="id">bioSources</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs29', 52)" onmouseover="showTip(event, 'fs29', 52)" class="fn">getRightShiftedBestPWMSs</span> <span onmouseout="hideTip(event, 'fs21', 53)" onmouseover="showTip(event, 'fs21', 53)" class="id">motiveLength</span> <span onmouseout="hideTip(event, 'fs22', 54)" onmouseover="showTip(event, 'fs22', 54)" class="id">pseudoCount</span> <span onmouseout="hideTip(event, 'fs14', 55)" onmouseover="showTip(event, 'fs14', 55)" class="id">alphabet</span> <span onmouseout="hideTip(event, 'fs6', 56)" onmouseover="showTip(event, 'fs6', 56)" class="id">bioSources</span>
</code></pre></td>
</tr>
</table>
<p>These functions are doing a side check, which means, that ich looks if upstream or downstream of the position a better sequence is, because one of the weakpoints of
the Gibbs Sampling Algorithm is, that it can have difficulty finding the global maximum due to local maxima. With the shift the function checks whether there are
better fitting motifs close to the motifs of the first convergence or not and if yes, gives them to you.</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l"> 1: </span>
<span class="l"> 2: </span>
<span class="l"> 3: </span>
<span class="l"> 4: </span>
<span class="l"> 5: </span>
<span class="l"> 6: </span>
<span class="l"> 7: </span>
<span class="l"> 8: </span>
<span class="l"> 9: </span>
<span class="l">10: </span>
<span class="l">11: </span>
<span class="l">12: </span>
<span class="l">13: </span>
<span class="l">14: </span>
<span class="l">15: </span>
<span class="l">16: </span>
<span class="l">17: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs30', 57)" onmouseover="showTip(event, 'fs30', 57)" class="id">backGroundProbs</span> <span class="o">=</span>
    <span onmouseout="hideTip(event, 'fs31', 58)" onmouseover="showTip(event, 'fs31', 58)" class="m">CompositeVector</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs32', 59)" onmouseover="showTip(event, 'fs32', 59)" class="id">createFCVOf</span> <span onmouseout="hideTip(event, 'fs14', 60)" onmouseover="showTip(event, 'fs14', 60)" class="id">alphabet</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs31', 61)" onmouseover="showTip(event, 'fs31', 61)" class="m">CompositeVector</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs33', 62)" onmouseover="showTip(event, 'fs33', 62)" class="id">createPCVOf</span>

<span class="k">let</span> <span onmouseout="hideTip(event, 'fs34', 63)" onmouseover="showTip(event, 'fs34', 63)" class="id">highThyminLevels</span> <span class="o">=</span>
    <span class="k">let</span> <span onmouseout="hideTip(event, 'fs35', 64)" onmouseover="showTip(event, 'fs35', 64)" class="id">tmp</span> <span class="o">=</span> <span class="k">new</span> <span onmouseout="hideTip(event, 'fs31', 65)" onmouseover="showTip(event, 'fs31', 65)" class="m">CompositeVector</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs36', 66)" onmouseover="showTip(event, 'fs36', 66)" class="id">ProbabilityCompositeVector</span><span class="pn">(</span><span class="pn">)</span>
    <span onmouseout="hideTip(event, 'fs31', 67)" onmouseover="showTip(event, 'fs31', 67)" class="m">CompositeVector</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs37', 68)" onmouseover="showTip(event, 'fs37', 68)" class="id">increaseInPlacePCVBy</span> <span onmouseout="hideTip(event, 'fs15', 69)" onmouseover="showTip(event, 'fs15', 69)" class="m">Nucleotides</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs18', 70)" onmouseover="showTip(event, 'fs18', 70)" class="id">T</span> <span class="n">0.7</span> <span onmouseout="hideTip(event, 'fs35', 71)" onmouseover="showTip(event, 'fs35', 71)" class="id">tmp</span> <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs38', 72)" onmouseover="showTip(event, 'fs38', 72)" class="fn">ignore</span>
    <span onmouseout="hideTip(event, 'fs31', 73)" onmouseover="showTip(event, 'fs31', 73)" class="m">CompositeVector</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs37', 74)" onmouseover="showTip(event, 'fs37', 74)" class="id">increaseInPlacePCVBy</span> <span onmouseout="hideTip(event, 'fs15', 75)" onmouseover="showTip(event, 'fs15', 75)" class="m">Nucleotides</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs17', 76)" onmouseover="showTip(event, 'fs17', 76)" class="id">A</span> <span class="n">0.1</span> <span onmouseout="hideTip(event, 'fs35', 77)" onmouseover="showTip(event, 'fs35', 77)" class="id">tmp</span> <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs38', 78)" onmouseover="showTip(event, 'fs38', 78)" class="fn">ignore</span>
    <span onmouseout="hideTip(event, 'fs31', 79)" onmouseover="showTip(event, 'fs31', 79)" class="m">CompositeVector</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs37', 80)" onmouseover="showTip(event, 'fs37', 80)" class="id">increaseInPlacePCVBy</span> <span onmouseout="hideTip(event, 'fs15', 81)" onmouseover="showTip(event, 'fs15', 81)" class="m">Nucleotides</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs19', 82)" onmouseover="showTip(event, 'fs19', 82)" class="id">G</span> <span class="n">0.1</span> <span onmouseout="hideTip(event, 'fs35', 83)" onmouseover="showTip(event, 'fs35', 83)" class="id">tmp</span> <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs38', 84)" onmouseover="showTip(event, 'fs38', 84)" class="fn">ignore</span>
    <span onmouseout="hideTip(event, 'fs31', 85)" onmouseover="showTip(event, 'fs31', 85)" class="m">CompositeVector</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs37', 86)" onmouseover="showTip(event, 'fs37', 86)" class="id">increaseInPlacePCVBy</span> <span onmouseout="hideTip(event, 'fs15', 87)" onmouseover="showTip(event, 'fs15', 87)" class="m">Nucleotides</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs20', 88)" onmouseover="showTip(event, 'fs20', 88)" class="id">C</span> <span class="n">0.1</span> <span onmouseout="hideTip(event, 'fs35', 89)" onmouseover="showTip(event, 'fs35', 89)" class="id">tmp</span> <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs38', 90)" onmouseover="showTip(event, 'fs38', 90)" class="fn">ignore</span>
    <span onmouseout="hideTip(event, 'fs35', 91)" onmouseover="showTip(event, 'fs35', 91)" class="id">tmp</span>

<span class="k">let</span> <span onmouseout="hideTip(event, 'fs39', 92)" onmouseover="showTip(event, 'fs39', 92)" class="id">bestFittingMotifsWithBackGroundKnowledge</span> <span class="o">=</span>
    <span onmouseout="hideTip(event, 'fs40', 93)" onmouseover="showTip(event, 'fs40', 93)" class="fn">getPWMOfRandomStartsWithBPV</span> <span onmouseout="hideTip(event, 'fs21', 94)" onmouseover="showTip(event, 'fs21', 94)" class="id">motiveLength</span> <span onmouseout="hideTip(event, 'fs22', 95)" onmouseover="showTip(event, 'fs22', 95)" class="id">pseudoCount</span> <span onmouseout="hideTip(event, 'fs14', 96)" onmouseover="showTip(event, 'fs14', 96)" class="id">alphabet</span> <span onmouseout="hideTip(event, 'fs6', 97)" onmouseover="showTip(event, 'fs6', 97)" class="id">bioSources</span> <span onmouseout="hideTip(event, 'fs30', 98)" onmouseover="showTip(event, 'fs30', 98)" class="id">backGroundProbs</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs41', 99)" onmouseover="showTip(event, 'fs41', 99)" class="fn">findBestmotiveWithStartPosition</span> <span onmouseout="hideTip(event, 'fs21', 100)" onmouseover="showTip(event, 'fs21', 100)" class="id">motiveLength</span> <span onmouseout="hideTip(event, 'fs22', 101)" onmouseover="showTip(event, 'fs22', 101)" class="id">pseudoCount</span> <span onmouseout="hideTip(event, 'fs14', 102)" onmouseover="showTip(event, 'fs14', 102)" class="id">alphabet</span> <span onmouseout="hideTip(event, 'fs6', 103)" onmouseover="showTip(event, 'fs6', 103)" class="id">bioSources</span> <span onmouseout="hideTip(event, 'fs30', 104)" onmouseover="showTip(event, 'fs30', 104)" class="id">backGroundProbs</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs42', 105)" onmouseover="showTip(event, 'fs42', 105)" class="fn">getLeftShiftedBestPWMSsWithBPV</span> <span onmouseout="hideTip(event, 'fs21', 106)" onmouseover="showTip(event, 'fs21', 106)" class="id">motiveLength</span> <span onmouseout="hideTip(event, 'fs22', 107)" onmouseover="showTip(event, 'fs22', 107)" class="id">pseudoCount</span> <span onmouseout="hideTip(event, 'fs14', 108)" onmouseover="showTip(event, 'fs14', 108)" class="id">alphabet</span> <span onmouseout="hideTip(event, 'fs6', 109)" onmouseover="showTip(event, 'fs6', 109)" class="id">bioSources</span> <span onmouseout="hideTip(event, 'fs30', 110)" onmouseover="showTip(event, 'fs30', 110)" class="id">backGroundProbs</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs43', 111)" onmouseover="showTip(event, 'fs43', 111)" class="fn">getRightShiftedBestPWMSsWithBPV</span> <span onmouseout="hideTip(event, 'fs21', 112)" onmouseover="showTip(event, 'fs21', 112)" class="id">motiveLength</span> <span onmouseout="hideTip(event, 'fs22', 113)" onmouseover="showTip(event, 'fs22', 113)" class="id">pseudoCount</span> <span onmouseout="hideTip(event, 'fs14', 114)" onmouseover="showTip(event, 'fs14', 114)" class="id">alphabet</span> <span onmouseout="hideTip(event, 'fs6', 115)" onmouseover="showTip(event, 'fs6', 115)" class="id">bioSources</span> <span onmouseout="hideTip(event, 'fs30', 116)" onmouseover="showTip(event, 'fs30', 116)" class="id">backGroundProbs</span>
</code></pre></td>
</tr>
</table>
<p>The combination of these functions can be used like the ones before but in contrast to those they require an externe Probability-Composite-Vector.
This is avantegous, because it enables you to create a specific background for the motif you are looking for and the usage of prior knowledge, e.g.
the genome has an equal distribution of all four bases, then you can just create a PCV based on the alphabet given in this documentation but an PCV
with the desired probabilities can easily be created otherwise too, using the increaseInPlace functions.</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l"> 1: </span>
<span class="l"> 2: </span>
<span class="l"> 3: </span>
<span class="l"> 4: </span>
<span class="l"> 5: </span>
<span class="l"> 6: </span>
<span class="l"> 7: </span>
<span class="l"> 8: </span>
<span class="l"> 9: </span>
<span class="l">10: </span>
<span class="l">11: </span>
<span class="l">12: </span>
<span class="l">13: </span>
<span class="l">14: </span>
<span class="l">15: </span>
<span class="l">16: </span>
<span class="l">17: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs44', 117)" onmouseover="showTip(event, 'fs44', 117)" class="id">ppm</span> <span class="o">=</span>
    <span class="pn">[|</span>
        <span class="s">&quot;CACGTG&quot;</span>
        <span class="s">&quot;CACGTG&quot;</span>
        <span class="s">&quot;CACGTG&quot;</span>
        <span class="s">&quot;CACGTG&quot;</span>
    <span class="pn">|]</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs12', 118)" onmouseover="showTip(event, 'fs12', 118)" class="m">Array</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs45', 119)" onmouseover="showTip(event, 'fs45', 119)" class="id">map</span> <span class="pn">(</span><span class="k">fun</span> <span onmouseout="hideTip(event, 'fs9', 120)" onmouseover="showTip(event, 'fs9', 120)" class="id">item</span> <span class="k">-&gt;</span> <span onmouseout="hideTip(event, 'fs10', 121)" onmouseover="showTip(event, 'fs10', 121)" class="m">BioArray</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs11', 122)" onmouseover="showTip(event, 'fs11', 122)" class="id">ofNucleotideString</span> <span onmouseout="hideTip(event, 'fs9', 123)" onmouseover="showTip(event, 'fs9', 123)" class="id">item</span><span class="pn">)</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs12', 124)" onmouseover="showTip(event, 'fs12', 124)" class="m">Array</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs45', 125)" onmouseover="showTip(event, 'fs45', 125)" class="id">map</span> <span class="pn">(</span><span class="k">fun</span> <span onmouseout="hideTip(event, 'fs46', 126)" onmouseover="showTip(event, 'fs46', 126)" class="id">item</span> <span class="k">-&gt;</span> <span onmouseout="hideTip(event, 'fs5', 127)" onmouseover="showTip(event, 'fs5', 127)" class="m">PositionMatrix</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs47', 128)" onmouseover="showTip(event, 'fs47', 128)" class="id">createPFMOf</span> <span onmouseout="hideTip(event, 'fs46', 129)" onmouseover="showTip(event, 'fs46', 129)" class="id">item</span><span class="pn">)</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs5', 130)" onmouseover="showTip(event, 'fs5', 130)" class="m">PositionMatrix</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs48', 131)" onmouseover="showTip(event, 'fs48', 131)" class="id">fusePositionFrequencyMatrices</span> <span onmouseout="hideTip(event, 'fs21', 132)" onmouseover="showTip(event, 'fs21', 132)" class="id">motiveLength</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs5', 133)" onmouseover="showTip(event, 'fs5', 133)" class="m">PositionMatrix</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs49', 134)" onmouseover="showTip(event, 'fs49', 134)" class="id">createPPMOf</span>

<span class="k">let</span> <span onmouseout="hideTip(event, 'fs50', 135)" onmouseover="showTip(event, 'fs50', 135)" class="id">bestFittingMotifsWithMotifKnowledge</span> <span class="o">=</span>
    <span onmouseout="hideTip(event, 'fs51', 136)" onmouseover="showTip(event, 'fs51', 136)" class="fn">getmotifsWithBestPWMSOfPPM</span> <span onmouseout="hideTip(event, 'fs21', 137)" onmouseover="showTip(event, 'fs21', 137)" class="id">motiveLength</span> <span onmouseout="hideTip(event, 'fs22', 138)" onmouseover="showTip(event, 'fs22', 138)" class="id">pseudoCount</span> <span onmouseout="hideTip(event, 'fs14', 139)" onmouseover="showTip(event, 'fs14', 139)" class="id">alphabet</span> <span onmouseout="hideTip(event, 'fs6', 140)" onmouseover="showTip(event, 'fs6', 140)" class="id">bioSources</span> <span onmouseout="hideTip(event, 'fs44', 141)" onmouseover="showTip(event, 'fs44', 141)" class="id">ppm</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs26', 142)" onmouseover="showTip(event, 'fs26', 142)" class="fn">getBestPWMSsWithStartPositions</span> <span onmouseout="hideTip(event, 'fs21', 143)" onmouseover="showTip(event, 'fs21', 143)" class="id">motiveLength</span> <span onmouseout="hideTip(event, 'fs22', 144)" onmouseover="showTip(event, 'fs22', 144)" class="id">pseudoCount</span> <span onmouseout="hideTip(event, 'fs14', 145)" onmouseover="showTip(event, 'fs14', 145)" class="id">alphabet</span> <span onmouseout="hideTip(event, 'fs6', 146)" onmouseover="showTip(event, 'fs6', 146)" class="id">bioSources</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs28', 147)" onmouseover="showTip(event, 'fs28', 147)" class="fn">getLeftShiftedBestPWMSs</span> <span onmouseout="hideTip(event, 'fs21', 148)" onmouseover="showTip(event, 'fs21', 148)" class="id">motiveLength</span> <span onmouseout="hideTip(event, 'fs22', 149)" onmouseover="showTip(event, 'fs22', 149)" class="id">pseudoCount</span> <span onmouseout="hideTip(event, 'fs14', 150)" onmouseover="showTip(event, 'fs14', 150)" class="id">alphabet</span> <span onmouseout="hideTip(event, 'fs6', 151)" onmouseover="showTip(event, 'fs6', 151)" class="id">bioSources</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs29', 152)" onmouseover="showTip(event, 'fs29', 152)" class="fn">getRightShiftedBestPWMSs</span> <span onmouseout="hideTip(event, 'fs21', 153)" onmouseover="showTip(event, 'fs21', 153)" class="id">motiveLength</span> <span onmouseout="hideTip(event, 'fs22', 154)" onmouseover="showTip(event, 'fs22', 154)" class="id">pseudoCount</span> <span onmouseout="hideTip(event, 'fs14', 155)" onmouseover="showTip(event, 'fs14', 155)" class="id">alphabet</span> <span onmouseout="hideTip(event, 'fs6', 156)" onmouseover="showTip(event, 'fs6', 156)" class="id">bioSources</span>
</code></pre></td>
</tr>
</table>
<p>These functions are best used when you already know the motif you are looking for but lack knowledge of the background. They skip the
random initlialization step and use the given PPM to find the best fitting motifs in the first iteration step and update it with every
found motif with a higher PWMS.</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l"> 1: </span>
<span class="l"> 2: </span>
<span class="l"> 3: </span>
<span class="l"> 4: </span>
<span class="l"> 5: </span>
<span class="l"> 6: </span>
<span class="l"> 7: </span>
<span class="l"> 8: </span>
<span class="l"> 9: </span>
<span class="l">10: </span>
<span class="l">11: </span>
<span class="l">12: </span>
<span class="l">13: </span>
<span class="l">14: </span>
<span class="l">15: </span>
<span class="l">16: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs52', 157)" onmouseover="showTip(event, 'fs52', 157)" class="id">pwm</span> <span class="o">=</span>
    <span class="k">let</span> <span onmouseout="hideTip(event, 'fs53', 158)" onmouseover="showTip(event, 'fs53', 158)" class="id">tmp</span> <span class="o">=</span> <span class="k">new</span> <span onmouseout="hideTip(event, 'fs54', 159)" onmouseover="showTip(event, 'fs54', 159)" class="rt">PositionWeightMatrix</span><span class="pn">(</span><span onmouseout="hideTip(event, 'fs21', 160)" onmouseover="showTip(event, 'fs21', 160)" class="id">motiveLength</span><span class="pn">)</span>
    <span class="pn">[|</span><span class="n">0..</span><span class="n">5</span><span class="pn">|]</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs12', 161)" onmouseover="showTip(event, 'fs12', 161)" class="m">Array</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs45', 162)" onmouseover="showTip(event, 'fs45', 162)" class="id">map</span> <span class="pn">(</span><span class="k">fun</span> <span onmouseout="hideTip(event, 'fs55', 163)" onmouseover="showTip(event, 'fs55', 163)" class="id">item</span> <span class="k">-&gt;</span>
        <span onmouseout="hideTip(event, 'fs5', 164)" onmouseover="showTip(event, 'fs5', 164)" class="m">PositionMatrix</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs56', 165)" onmouseover="showTip(event, 'fs56', 165)" class="id">increaseInPlacePWMBy</span> <span onmouseout="hideTip(event, 'fs15', 166)" onmouseover="showTip(event, 'fs15', 166)" class="m">Nucleotides</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs18', 167)" onmouseover="showTip(event, 'fs18', 167)" class="id">T</span> <span onmouseout="hideTip(event, 'fs55', 168)" onmouseover="showTip(event, 'fs55', 168)" class="id">item</span> <span class="n">0.7</span> <span onmouseout="hideTip(event, 'fs53', 169)" onmouseover="showTip(event, 'fs53', 169)" class="id">tmp</span> <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs38', 170)" onmouseover="showTip(event, 'fs38', 170)" class="fn">ignore</span>
        <span onmouseout="hideTip(event, 'fs5', 171)" onmouseover="showTip(event, 'fs5', 171)" class="m">PositionMatrix</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs56', 172)" onmouseover="showTip(event, 'fs56', 172)" class="id">increaseInPlacePWMBy</span> <span onmouseout="hideTip(event, 'fs15', 173)" onmouseover="showTip(event, 'fs15', 173)" class="m">Nucleotides</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs17', 174)" onmouseover="showTip(event, 'fs17', 174)" class="id">A</span> <span onmouseout="hideTip(event, 'fs55', 175)" onmouseover="showTip(event, 'fs55', 175)" class="id">item</span> <span class="n">0.1</span> <span onmouseout="hideTip(event, 'fs53', 176)" onmouseover="showTip(event, 'fs53', 176)" class="id">tmp</span> <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs38', 177)" onmouseover="showTip(event, 'fs38', 177)" class="fn">ignore</span>
        <span onmouseout="hideTip(event, 'fs5', 178)" onmouseover="showTip(event, 'fs5', 178)" class="m">PositionMatrix</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs56', 179)" onmouseover="showTip(event, 'fs56', 179)" class="id">increaseInPlacePWMBy</span> <span onmouseout="hideTip(event, 'fs15', 180)" onmouseover="showTip(event, 'fs15', 180)" class="m">Nucleotides</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs19', 181)" onmouseover="showTip(event, 'fs19', 181)" class="id">G</span> <span onmouseout="hideTip(event, 'fs55', 182)" onmouseover="showTip(event, 'fs55', 182)" class="id">item</span> <span class="n">0.1</span> <span onmouseout="hideTip(event, 'fs53', 183)" onmouseover="showTip(event, 'fs53', 183)" class="id">tmp</span> <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs38', 184)" onmouseover="showTip(event, 'fs38', 184)" class="fn">ignore</span>
        <span onmouseout="hideTip(event, 'fs5', 185)" onmouseover="showTip(event, 'fs5', 185)" class="m">PositionMatrix</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs56', 186)" onmouseover="showTip(event, 'fs56', 186)" class="id">increaseInPlacePWMBy</span> <span onmouseout="hideTip(event, 'fs15', 187)" onmouseover="showTip(event, 'fs15', 187)" class="m">Nucleotides</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs20', 188)" onmouseover="showTip(event, 'fs20', 188)" class="id">C</span> <span onmouseout="hideTip(event, 'fs55', 189)" onmouseover="showTip(event, 'fs55', 189)" class="id">item</span> <span class="n">0.1</span> <span onmouseout="hideTip(event, 'fs53', 190)" onmouseover="showTip(event, 'fs53', 190)" class="id">tmp</span> <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs38', 191)" onmouseover="showTip(event, 'fs38', 191)" class="fn">ignore</span>
                <span class="pn">)</span> <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs38', 192)" onmouseover="showTip(event, 'fs38', 192)" class="fn">ignore</span>
    <span onmouseout="hideTip(event, 'fs53', 193)" onmouseover="showTip(event, 'fs53', 193)" class="id">tmp</span>

<span class="k">let</span> <span onmouseout="hideTip(event, 'fs57', 194)" onmouseover="showTip(event, 'fs57', 194)" class="id">bestFittingMotifsWithGivenPWM</span> <span class="o">=</span>
    <span onmouseout="hideTip(event, 'fs58', 195)" onmouseover="showTip(event, 'fs58', 195)" class="fn">getmotifsWithBestPWMSOfPWM</span> <span onmouseout="hideTip(event, 'fs21', 196)" onmouseover="showTip(event, 'fs21', 196)" class="id">motiveLength</span> <span onmouseout="hideTip(event, 'fs6', 197)" onmouseover="showTip(event, 'fs6', 197)" class="id">bioSources</span> <span onmouseout="hideTip(event, 'fs52', 198)" onmouseover="showTip(event, 'fs52', 198)" class="id">pwm</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs26', 199)" onmouseover="showTip(event, 'fs26', 199)" class="fn">getBestPWMSsWithStartPositions</span> <span onmouseout="hideTip(event, 'fs21', 200)" onmouseover="showTip(event, 'fs21', 200)" class="id">motiveLength</span> <span onmouseout="hideTip(event, 'fs22', 201)" onmouseover="showTip(event, 'fs22', 201)" class="id">pseudoCount</span> <span onmouseout="hideTip(event, 'fs14', 202)" onmouseover="showTip(event, 'fs14', 202)" class="id">alphabet</span> <span onmouseout="hideTip(event, 'fs6', 203)" onmouseover="showTip(event, 'fs6', 203)" class="id">bioSources</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs28', 204)" onmouseover="showTip(event, 'fs28', 204)" class="fn">getLeftShiftedBestPWMSs</span> <span onmouseout="hideTip(event, 'fs21', 205)" onmouseover="showTip(event, 'fs21', 205)" class="id">motiveLength</span> <span onmouseout="hideTip(event, 'fs22', 206)" onmouseover="showTip(event, 'fs22', 206)" class="id">pseudoCount</span> <span onmouseout="hideTip(event, 'fs14', 207)" onmouseover="showTip(event, 'fs14', 207)" class="id">alphabet</span> <span onmouseout="hideTip(event, 'fs6', 208)" onmouseover="showTip(event, 'fs6', 208)" class="id">bioSources</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs29', 209)" onmouseover="showTip(event, 'fs29', 209)" class="fn">getRightShiftedBestPWMSs</span> <span onmouseout="hideTip(event, 'fs21', 210)" onmouseover="showTip(event, 'fs21', 210)" class="id">motiveLength</span> <span onmouseout="hideTip(event, 'fs22', 211)" onmouseover="showTip(event, 'fs22', 211)" class="id">pseudoCount</span> <span onmouseout="hideTip(event, 'fs14', 212)" onmouseover="showTip(event, 'fs14', 212)" class="id">alphabet</span> <span onmouseout="hideTip(event, 'fs6', 213)" onmouseover="showTip(event, 'fs6', 213)" class="id">bioSources</span>
</code></pre></td>
</tr>
</table>
<p>The getmotifsWithBestPWMSOfPWM finds the best motif for every sequence based on the given PWM and without a randwom start. The PWM is only
updated after it has found the best motif for every squence. Based on the information available to you, you can change the functions
that follwo the getmotifsWithBestPWMSOfPWM with one of teh above functions.</p>
<p>Gibbs-Sampling is an algorithm that can be used with little prio knowledge to find conserved motifs in sequences by stumpling into one motif
at random, which biases the matrix to find more similiar motifs. Even when it can be used with little prior knowledge its sensitivity and speed
can be increased and randomness reduced with more knowledge. But still, the alggorithm has the problem of being heuristic and not finding the
same motifs and so several repetitions and result comparrisons are needed. Also can local maxima be a problem but their burden can be reduced,
thanks to the shift functions.
The Gibbs-Sampling-Algorithm is a useful tool to find conserved structures in biological sequences because it is easy to use and besides its heuristic
nature, it works reliable.
Thank you for reading this long post! Good luck with your motif search!</p>

<div class="tip" id="fs1">namespace BioFSharp</div>
<div class="tip" id="fs2">namespace FSharpAux</div>
<div class="tip" id="fs3">namespace GibbsSampling</div>
<div class="tip" id="fs4">module SiteSampler<br /><br />from GibbsSampling</div>
<div class="tip" id="fs5">module PositionMatrix<br /><br />from GibbsSampling</div>
<div class="tip" id="fs6">val bioSources : BioArray.BioArray&lt;Nucleotides.Nucleotide&gt; []</div>
<div class="tip" id="fs7">Multiple items<br />module Seq<br /><br />from FSharpAux<br /><br />--------------------<br />module Seq<br /><br />from Microsoft.FSharp.Collections</div>
<div class="tip" id="fs8">val map : mapping:(&#39;T -&gt; &#39;U) -&gt; source:seq&lt;&#39;T&gt; -&gt; seq&lt;&#39;U&gt;</div>
<div class="tip" id="fs9">val item : string</div>
<div class="tip" id="fs10">module BioArray<br /><br />from BioFSharp</div>
<div class="tip" id="fs11">val ofNucleotideString : s:#seq&lt;char&gt; -&gt; BioArray.BioArray&lt;Nucleotides.Nucleotide&gt;</div>
<div class="tip" id="fs12">Multiple items<br />module Array<br /><br />from FSharpAux<br /><br />--------------------<br />module Array<br /><br />from Microsoft.FSharp.Collections</div>
<div class="tip" id="fs13">val ofSeq : source:seq&lt;&#39;T&gt; -&gt; &#39;T []</div>
<div class="tip" id="fs14">val alphabet : Nucleotides.Nucleotide []</div>
<div class="tip" id="fs15">module Nucleotides<br /><br />from BioFSharp</div>
<div class="tip" id="fs16">type Nucleotide =<br />&#160;&#160;| A<br />&#160;&#160;| T<br />&#160;&#160;| G<br />&#160;&#160;| C<br />&#160;&#160;| U<br />&#160;&#160;| I<br />&#160;&#160;| Gap<br />&#160;&#160;| Ter<br />&#160;&#160;| R<br />&#160;&#160;| Y<br />&#160;&#160;...<br />&#160;&#160;&#160;&#160;interface IBioItem<br />&#160;&#160;&#160;&#160;static member op_Explicit : value:#IBioItem -&gt; int<br />&#160;&#160;&#160;&#160;static member op_Explicit : value:#IBioItem -&gt; byte</div>
<div class="tip" id="fs17">union case Nucleotides.Nucleotide.A: Nucleotides.Nucleotide</div>
<div class="tip" id="fs18">union case Nucleotides.Nucleotide.T: Nucleotides.Nucleotide</div>
<div class="tip" id="fs19">union case Nucleotides.Nucleotide.G: Nucleotides.Nucleotide</div>
<div class="tip" id="fs20">union case Nucleotides.Nucleotide.C: Nucleotides.Nucleotide</div>
<div class="tip" id="fs21">val motiveLength : int</div>
<div class="tip" id="fs22">val pseudoCount : float</div>
<div class="tip" id="fs23">val randomMotifs : (float * int) []</div>
<div class="tip" id="fs24">val getPWMOfRandomStarts : motiveLength:int -&gt; pseudoCount:float -&gt; alphabet:#IBioItem [] -&gt; sources:BioArray.BioArray&lt;#IBioItem&gt; [] -&gt; (float * int) []</div>
<div class="tip" id="fs25">val firstConvergedMotifs : (float * int) []</div>
<div class="tip" id="fs26">val getBestPWMSsWithStartPositions : motiveLength:int -&gt; pseudoCount:float -&gt; alphabet:#IBioItem [] -&gt; sources:BioArray.BioArray&lt;#IBioItem&gt; [] -&gt; startPositions:(float * int) [] -&gt; (float * int) []</div>
<div class="tip" id="fs27">val bestFittingMotifs : (float * int) []</div>
<div class="tip" id="fs28">val getLeftShiftedBestPWMSs : motiveLength:int -&gt; pseudoCount:float -&gt; alphabet:#IBioItem [] -&gt; sources:BioArray.BioArray&lt;#IBioItem&gt; [] -&gt; startPositions:(float * int) [] -&gt; (float * int) []</div>
<div class="tip" id="fs29">val getRightShiftedBestPWMSs : motiveLength:int -&gt; pseudoCount:float -&gt; alphabet:#IBioItem [] -&gt; sources:BioArray.BioArray&lt;#IBioItem&gt; [] -&gt; startPositions:(float * int) [] -&gt; (float * int) []</div>
<div class="tip" id="fs30">val backGroundProbs : CompositeVector.ProbabilityCompositeVector</div>
<div class="tip" id="fs31">module CompositeVector<br /><br />from GibbsSampling</div>
<div class="tip" id="fs32">val createFCVOf : resSources:BioArray.BioArray&lt;#IBioItem&gt; -&gt; CompositeVector.FrequencyCompositeVector</div>
<div class="tip" id="fs33">val createPCVOf : caArray:CompositeVector.FrequencyCompositeVector -&gt; CompositeVector.ProbabilityCompositeVector</div>
<div class="tip" id="fs34">val highThyminLevels : CompositeVector.ProbabilityCompositeVector</div>
<div class="tip" id="fs35">val tmp : CompositeVector.ProbabilityCompositeVector</div>
<div class="tip" id="fs36">Multiple items<br />type ProbabilityCompositeVector =<br />&#160;&#160;inherit CompositeVector&lt;IBioItem,float&gt;<br />&#160;&#160;new : unit -&gt; ProbabilityCompositeVector<br />&#160;&#160;private new : array:float [] -&gt; ProbabilityCompositeVector<br /><br />--------------------<br />new : unit -&gt; CompositeVector.ProbabilityCompositeVector</div>
<div class="tip" id="fs37">val increaseInPlacePCVBy : bioItem:#IBioItem -&gt; n:float -&gt; backGroundProbabilityVector:CompositeVector.ProbabilityCompositeVector -&gt; CompositeVector.ProbabilityCompositeVector</div>
<div class="tip" id="fs38">val ignore : value:&#39;T -&gt; unit</div>
<div class="tip" id="fs39">val bestFittingMotifsWithBackGroundKnowledge : (float * int) []</div>
<div class="tip" id="fs40">val getPWMOfRandomStartsWithBPV : motiveLength:int -&gt; pseudoCount:float -&gt; alphabet:#IBioItem [] -&gt; sources:BioArray.BioArray&lt;#IBioItem&gt; [] -&gt; pcv:CompositeVector.ProbabilityCompositeVector -&gt; (float * int) []</div>
<div class="tip" id="fs41">val findBestmotiveWithStartPosition : motiveLength:int -&gt; pseudoCount:float -&gt; alphabet:#IBioItem [] -&gt; sources:BioArray.BioArray&lt;#IBioItem&gt; [] -&gt; pcv:CompositeVector.ProbabilityCompositeVector -&gt; startPositions:(float * int) [] -&gt; (float * int) []</div>
<div class="tip" id="fs42">val getLeftShiftedBestPWMSsWithBPV : motiveLength:int -&gt; pseudoCount:float -&gt; alphabet:#IBioItem [] -&gt; sources:BioArray.BioArray&lt;#IBioItem&gt; [] -&gt; pcv:CompositeVector.ProbabilityCompositeVector -&gt; startPositions:(float * int) [] -&gt; (float * int) []</div>
<div class="tip" id="fs43">val getRightShiftedBestPWMSsWithBPV : motiveLength:int -&gt; pseudoCount:float -&gt; alphabet:#IBioItem [] -&gt; sources:BioArray.BioArray&lt;#IBioItem&gt; [] -&gt; pcv:CompositeVector.ProbabilityCompositeVector -&gt; startPositions:(float * int) [] -&gt; (float * int) []</div>
<div class="tip" id="fs44">val ppm : PositionProbabilityMatrix</div>
<div class="tip" id="fs45">val map : mapping:(&#39;T -&gt; &#39;U) -&gt; array:&#39;T [] -&gt; &#39;U []</div>
<div class="tip" id="fs46">val item : BioArray.BioArray&lt;Nucleotides.Nucleotide&gt;</div>
<div class="tip" id="fs47">val createPFMOf : source:BioArray.BioArray&lt;#IBioItem&gt; -&gt; PositionFrequencyMatrix</div>
<div class="tip" id="fs48">val fusePositionFrequencyMatrices : motiveLength:int -&gt; countMatrices:PositionFrequencyMatrix [] -&gt; PositionFrequencyMatrix</div>
<div class="tip" id="fs49">val createPPMOf : positionFrequencyMatrix:PositionFrequencyMatrix -&gt; PositionProbabilityMatrix</div>
<div class="tip" id="fs50">val bestFittingMotifsWithMotifKnowledge : (float * int) []</div>
<div class="tip" id="fs51">val getmotifsWithBestPWMSOfPPM : motiveLength:int -&gt; pseudoCount:float -&gt; alphabet:#IBioItem [] -&gt; sources:BioArray.BioArray&lt;#IBioItem&gt; [] -&gt; positionProbabilityMatrix:PositionProbabilityMatrix -&gt; (float * int) []</div>
<div class="tip" id="fs52">val pwm : PositionWeightMatrix</div>
<div class="tip" id="fs53">val tmp : PositionWeightMatrix</div>
<div class="tip" id="fs54">Multiple items<br />type PositionWeightMatrix =<br />&#160;&#160;inherit BaseMatrix&lt;IBioItem,float&gt;<br />&#160;&#160;new : rowLength:int -&gt; PositionWeightMatrix<br />&#160;&#160;private new : matrix:float [,] -&gt; PositionWeightMatrix<br /><br />--------------------<br />new : rowLength:int -&gt; PositionWeightMatrix</div>
<div class="tip" id="fs55">val item : int</div>
<div class="tip" id="fs56">val increaseInPlacePWMBy : bioItem:#IBioItem -&gt; pos:int -&gt; n:float -&gt; positionWeightMatrix:PositionWeightMatrix -&gt; PositionWeightMatrix</div>
<div class="tip" id="fs57">val bestFittingMotifsWithGivenPWM : (float * int) []</div>
<div class="tip" id="fs58">val getmotifsWithBestPWMSOfPWM : motiveLength:int -&gt; sources:BioArray.BioArray&lt;#IBioItem&gt; [] -&gt; pwm:PositionWeightMatrix -&gt; (float * int) []</div>

        </div>
        <div class="span3">
          <img src="/GibbsSampler/img/logo.png" alt="F# Project" style="width:150px;margin:10px" />  
          <ul class="nav nav-list" id="menu" style="margin-top: 20px;">
              <li class="nav-header">GibbsSampler</li>
              
              <li class="divider"></li>
              
              <li><a href="https://github.com/Etschbeijer/GibbsSampler">Source Code on GitHub</a></li>
              <li><a href="/GibbsSampler/Documentation.html">Blog-Post</a></li>
              <li><a href="/GibbsSampler/license.html">License</a></li>
              <li><a href="/GibbsSampler/release-notes.html">Release Notes</a></li>

              
              

              <li class="nav-header">Documentation</li>
              <li><a href="/GibbsSampler/reference/index.html">API Reference</a></li>
          </ul>
        </div>
      </div>
    </div>
    <a href="https://github.com/Etschbeijer/GibbsSampler"><img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_gray_6d6d6d.png" alt="Fork me on GitHub"/></a>
  </body>
  </html>
